### PR TITLE
AVRO-3384: Define C# Coding Style Guidelines

### PR DESCRIPTION
**Pull request by KyleSchoonover**
Documenting the C# styling guidelines as they are implemented in the .editorconfig

### Jira

- [ ] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-3384

### Tests

- [ ] My PR does not need testing for this extremely good reason: documentation

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by KyleSchoonover** [comment]
@zcsizmadia @martin-g @RyanSkraba This is just an initial draft, but is this what you are looking for?  And by no means is this complete.

**Comment by zcsizmadia** [comment]
Niiiice!

**Comment by zcsizmadia** [comment]
@KyleSchoonover Do you think we should include the non-formatting related coding styles and the preferred usage of language features in this file? IMO that might make sense

**Comment by KyleSchoonover** [comment]
I was debating the same thing.  Let me think of reasons not to and if I don't come up with a good reason than yes.

**Comment by zcsizmadia** [comment]
I really do like the way you tweak the editorconfig and having the styling md file.

**Comment by KyleSchoonover** [comment]
@zcsizmadia I would call this the first stab.  There are some additional miscellaneous things that I want to add, but this should allow us to quickly review a lot of the rules that we have the options to configure.

**Comment by zcsizmadia** [comment]
It looks really nice:

1. How much is it aligned as of now with most of the formatting coding style?
2. I saw somewhere that `file_header_template` can be defined by a placeholder file. If that is possible it would be really sweet to have the license info in that file. I think the new file in VSCode honors it, so when you create a file, the header+template is added automatically, but not 100% sure about that

**Comment by KyleSchoonover** [comment]
1. We've still got a bunch of fixes, but all in all it's fairly minor since I started it.
2. I was actually planning on doing that.  Just want to make sure it formats like how we want.

**Comment by KyleSchoonover** [comment]
Also, I'm thinking Code Quality should be it's own thing, since it would involve implementing Code Analysis rules again.  And we can call the rules out in the .editorconfig

**Comment by zcsizmadia** [comment]
There are 2k+ warnings and messages caused by Intellisense (~~30 different compiler codes). IMO the next step in this ticket should be:

1. Identify which message codes should be ignored and which should be fixed
2. Identify the compiler messages which should be turned on (currently off) if any
3. Prioritize the compiler message codes selected in #1 and #2. Complexity, risk of breaking change, potential performance, better documentation, readablity  and some others should be the factors when deciding about the priorities.

Discussing the items in #1 IMO is important, because a balance has to be found between having a mature codebase, personal preferences and best practices.

Since there are potentially a lot of code changes, a good initial strategy should be defined how the changes (PRs) are made (e.g. to ensure the review process is as easy as possible)

I dont think any of these changes are a amust have in any of the upcoming releases (unless some bugs are discovered during the changes and code reviews).

**Comment by KyleSchoonover** [comment]
Here's your list of things to fix.
`IDE0001:Name can be simplified.										74`
`IDE0002:Name can be simplified.										117`
`IDE0003:Name can be simplified.										262`
`IDE0004:Cast is redundant.											29`
`IDE0005:Using directive is unnecessary.								39`
`IDE0017:Object initialization can be simplified							44`
`IDE0018:Variable declaration can be inlined								42`
`IDE0019:Use pattern matching										46`
`IDE0020:Use pattern matching										6`
`IDE0028:Collection initialization can be simplified						15`
`IDE0029:Null check can be simplified									4`
`IDE0034:'default' expression can be simplified							2`
`IDE0036:Modifiers are not ordered									2`
`IDE0038:Use pattern matching										12`
`IDE0040:Accessibility modifiers required								35`
`IDE0041:Null check can be simplified									6`
`IDE0044:Make field readonly											48`
`IDE0045:'if' statement can be simplified								19`
`IDE0046:'if' statement can be simplified								100`
`IDE0049:Name can be simplified.										51`
`IDE0055:Fix formatting												1054`
`IDE0059:Unnecessary assignment of a value to 'enumType'					2`
`IDE0059:Unnecessary assignment of a value to 'obj'						2`
`IDE0059:Unnecessary assignment of a value to 'rc'						3`
`IDE0059:Unnecessary assignment of a value to 'reader'					2`
`IDE0059:Unnecessary assignment of a value to 'recordtype'				2`
`IDE0059:Unnecessary assignment of a value to 's'						4`
`IDE0059:Unnecessary assignment of a value to 'schema'					2`
`IDE0066:Use 'switch' expression										12`
`IDE0071:Interpolation can be simplified								10`
`IDE0076:Invalid or missing target for 'SuppressMessageAttribute'			2`
`IDE0078:Use pattern matching										2`
`IDE0079:Remove unnecessary suppression								52`

**Comment by zcsizmadia** [comment]
My suggestion would be that PRs are grouped around the compiler warning. Eg. 1 PR fixes 1 particular compiler warning. In that case thereviewer has an easier job, just focusing one kind of change. If the PR is "too big" it might have to be split into multiple ones. I am not sure what is a too big PR btw.

Compiler codes to fix:
IDE0001
IDE0002
IDE0003
IDE0004
IDE0005
IDE0034
IDE0036
IDE0040
IDE0044
IDE0049
IDE0055
IDE0059
IDE0071
IDE0076
IDE0078
IDE0079

Compiler codes to turn off  for now:
IDE0017
IDE0018
IDE0019
IDE0028
IDE0029
IDE0041
IDE0038
IDE0045
IDE0046
IDE0066

In the PRs the main objective should be that it is as reviewer friendly is possible, so the diff is as easily readable as possible.

**Comment by zcsizmadia** [comment]
@KyleSchoonover I think the use of expression-bodied members should be discussed as well. 

My suggestion would be to wait with any style related PRs until we discuss the guidelines first.

**Comment by zcsizmadia** [comment]
That language feature is not supported, unless we override the language version in the csproj file. which IMO is not ideal

**Comment by KyleSchoonover** [comment]
This is what I have left in my local.  Let me know when we are ready to move forward.  I have everything up to date and ready for merge requests.

CS0618'SchemaNormalization.Empty64' is obsolete: 'This will be removed from the public API in a future version.'	6
IDE0001Name can be simplified.	2
IDE0002Name can be simplified.	1
IDE0004Cast is redundant.	2
IDE0036Modifiers are not ordered	1
IDE0045'if' statement can be simplified	4
IDE0046'if' statement can be simplified	4
IDE0054Use compound assignment	1
IDE0055Fix formatting	11
IDE0075Conditional expression can be simplified	6
IDE1006Naming rule violation: Missing prefix: 'I'	11
IDE1006Naming rule violation: These words must begin with upper case characters: addName	1
IDE1006Naming rule violation: These words must begin with upper case characters: addNamespace	1
IDE1006Naming rule violation: These words must begin with upper case characters: areEqual	3
IDE1006Naming rule violation: These words must begin with upper case characters: createDocComment	1
IDE1006Naming rule violation: These words must begin with upper case characters: createSchemaField	1
IDE1006Naming rule violation: These words must begin with upper case characters: findBranch	1
IDE1006Naming rule violation: These words must begin with upper case characters: generateNames	2
IDE1006Naming rule violation: These words must begin with upper case characters: getHashCode	2
IDE1006Naming rule violation: These words must begin with upper case characters: getNullableType	1
IDE1006Naming rule violation: These words must begin with upper case characters: getTargetType	1
IDE1006Naming rule violation: These words must begin with upper case characters: getType	1
IDE1006Naming rule violation: These words must begin with upper case characters: processEnum	1
IDE1006Naming rule violation: These words must begin with upper case characters: processFixed	1
IDE1006Naming rule violation: These words must begin with upper case characters: processInterface	1
IDE1006Naming rule violation: These words must begin with upper case characters: processProtocols	1
IDE1006Naming rule violation: These words must begin with upper case characters: processRecord	1
IDE1006Naming rule violation: These words must begin with upper case characters: processSchemas	1
IDE1006Naming rule violation: These words must begin with upper case characters: value	1
IDE1006Naming rule violation: These words must begin with upper case characters: writeIfNotNull	1
IDE1006Naming rule violation: These words must begin with upper case characters: writeIfNotNullOrEmpty	1
IDE1006Naming rule violation: These words must begin with upper case characters: writeJson	2
Grand Total	75

**Comment by zcsizmadia** [comment]
1. The pending PRs still have if statement changes, which I have concern about, since it complicates the review process
2. Expression bodied members are used for ne readablity gain. As I mentioned earlier we should discuss basic guidelines about that
3. I dont think the variables were supposed to be renamed, which causes the diffs being more complex.
4. ...

As I mentioned eearlier, createing the PRs should have been started after the guidelines are set and we agree how we proceed. This would make sure that we respect your time and the reviewers time as well. The concern is not that it cannot be reviewed, but the volume of the review will be significant.

**Comment by KyleSchoonover** [comment]
As for 1, 2 and 3 they are all defined in the editorconfig. Additionally, I only changed private name, which should affect nothing on the public API.  To implement these styling guidelines I find it best to just get it over with.  Otherwise you take something that can be done in a matter of days and drag it out for months.  The local tests pass, as well as this code has been ported over to my companies version of the library and I get the same results on over 6000 schemas.  These are all non breaking changes, and do not change the functionality of the library.  

I find it best to treat these as if they were completely new files.  Which should make those who review them more involved and informed what the code base does.

**Comment by KyleSchoonover** [comment]
Just for reference

- [Google C# Style guidelines](https://google.github.io/styleguide/csharp-style.html)
- [Dotnet runtime Style guidelines](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md)

I'm going to consider these "mature" styling practices.

**Comment by zcsizmadia** [comment]
What is the current state of the Intellisense warning and info messages? If there are still warnings and info messages, are you planning to leave them like that or override them in the styling definition files and address them in different tickets? e.g. each ticket for each code warning?

**Comment by KyleSchoonover** [comment]
> What is the current state of the Intellisense warning and info messages? If there are still warnings and info messages, are you planning to leave them like that or override them in the styling definition files and address them in different tickets? e.g. each ticket for each code warning?

Let's talk about implementation on slack.  This is just documentation and additions to the editorconfig.

**Comment by KyleSchoonover** [comment]
As far as I can tell this is just documenting what we've put in editorconfig.  If we want to talk about going to implement all these rules, I would rather tasks and conversation happen here: https://issues.apache.org/jira/browse/AVRO-3383.

**Review by zcsizmadia — APPROVED** [review]

File: lang/csharp/STYLING.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
+# C# Styling Rules for Apache.Avro
+
+The following rules are currently used within the .editorconfig of the Avro solution.  Any changes to this documentation should be reflected in the .editorconfig file and vice versa.
```
[the remark this answers is not in the document]
Comment (KyleSchoonover) [reply lang/csharp/STYLING.md:22]: Added it to the notes section

**Comment by martin-g** [comment]
Thank you, @KyleSchoonover !

**Pull request merged**